File: rust/src/code/longest_substring_without_repeating_characters.rs
use std::{cmp, collections::HashSet};

pub struct Solution {}

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut max_len = 0;
        let mut l = 0;
        let mut r = 0;
        let mut set: HashSet<char> = HashSet::new();
        let chars: Vec<char> = s.chars().collect();

        while r < chars.len() {
            if set.contains(&chars[r]) {
                set.remove(&chars[l]);
                l += 1;
            } else {
                set.insert(chars[r]);
                max_len = cmp::max(max_len, set.len());
                r += 1;
            }
        }

        return max_len as i32;
    }

    pub fn length_of_longest_substring_2(s: String) -> i32 {
        let mut l = 0;
        let mut max_len = 0;
        let mut set: HashSet<char> = HashSet::new();
        let chars: Vec<char> = s.chars().collect();

        for r in 0..chars.len() {
            while set.contains(&chars[r]) {
                set.remove(&chars[l]);
                l += 1;
            }

            set.insert(chars[r]);
            max_len = cmp::max(max_len, set.len());
        }

        return max_len as i32;
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn length_of_longest_substring_case_1() {
        // arrange
        let s = String::from("abcabcbb");
        let expected = 3;

        // act
        let actual = Solution::length_of_longest_substring(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn length_of_longest_substring_case_2() {
        // arrange
        let s = String::from("bbbbb");
        let expected = 1;

        // act
        let actual = Solution::length_of_longest_substring(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn length_of_longest_substring_case_3() {
        // arrange
        let s = String::from("pwwkew");
        let expected3 = 3;

        // act
        let actual = Solution::length_of_longest_substring(s);

        // assert
        assert_eq!(actual, expected3);
    }

    #[test]
    fn length_of_longest_substring_case_4() {
        // arrange
        let s = String::from("aabaab!bb");
        let expected = 3;

        // act
        let actual = Solution::length_of_longest_substring(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn length_of_longest_substring_case_5() {
        // arrange
        let s = String::from("abcabcbb");
        let expected = 3;

        // act
        let actual = Solution::length_of_longest_substring_2(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn length_of_longest_substring_case_6() {
        // arrange
        let s = String::from("bbbbb");
        let expected = 1;

        // act
        let actual = Solution::length_of_longest_substring_2(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn length_of_longest_substring_case_7() {
        // arrange
        let s = String::from("pwwkew");
        let expected = 3;

        // act
        let actual = Solution::length_of_longest_substring_2(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn length_of_longest_substring_case_8() {
        // arrange
        let s = String::from("aabaab!bb");
        let expected = 3;

        // act
        let actual = Solution::length_of_longest_substring_2(s);

        // assert
        assert_eq!(actual, expected);
    }
}

File: python/longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """time complexity: O(n), space complexity: O(n)"""
        result: int = 0
        char_set: set[str] = set()
        l: int = 0
        for r in range(len(s)):
            while l < r and s[r] in char_set:
                char_set.remove(s[l])
                l += 1
            char_set.add(s[r])
            result = max(result, r - l + 1)

        return result


def test_lengthOfLongestSubstring_case_1():
    # arrange
    s: str = "abcabcbb"
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.lengthOfLongestSubstring(s)

    # assert
    assert expected == actual


def test_lengthOfLongestSubstring_case_2():
    # arrange
    s: str = "bbbb"
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.lengthOfLongestSubstring(s)

    # assert
    assert expected == actual


def test_lengthOfLongestSubstring_case_3():
    # arrange
    s: str = "pwwkew"
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.lengthOfLongestSubstring(s)

    # assert
    assert expected == actual
